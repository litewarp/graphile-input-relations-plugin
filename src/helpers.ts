import type {
  PgCodecRelation,
  PgCodecWithAttributes,
  PgRegistry,
  PgResource,
  PgResourceUnique,
} from 'postgraphile/@dataplan/pg';
import type {} from 'postgraphile/graphile-build-pg';

export function isPgTableResource(r: PgResource): r is PgTableResource {
  return Boolean(r.codec.attributes) && !r.parameters;
}

export interface PgCodecRelationWithName
  extends PgCodecRelation<PgCodecWithAttributes, PgTableResource> {
  name: string;
  resource: string;
}
export type PgTableResource = PgResource<
  string,
  PgCodecWithAttributes,
  PgResourceUnique[],
  undefined,
  PgRegistry
>;

export const isNestedMutableResource = (
  build: GraphileBuild.Build,
  resource: PgResource
): resource is PgTableResource => {
  if (resource.parameters) return false;
  if (!resource.codec.attributes) return false;
  if (resource.codec.polymorphism) return false;
  if (resource.codec.isAnonymous) return false;
  const behaviors = ['resource:insert', 'resource:update', 'resource:delete'] as const;
  return behaviors.some((behavior) =>
    build.behavior.pgResourceMatches(resource, behavior)
  );
};

export const isInsertable = (build: GraphileBuild.Build, resource: PgTableResource) => {
  if (resource.parameters) return false;
  if (!resource.codec.attributes) return false;
  if (resource.codec.polymorphism) return false;
  if (resource.codec.isAnonymous) return false;
  return build.behavior.pgResourceMatches(resource, 'resource:insert') === true;
};

export const isUpdatable = (build: GraphileBuild.Build, resource: PgTableResource) => {
  if (resource.parameters) return false;
  if (!resource.codec.attributes) return false;
  if (resource.codec.polymorphism) return false;
  if (resource.codec.isAnonymous) return false;
  if (!resource.uniques || resource.uniques.length < 1) return false;
  return Boolean(build.behavior.pgResourceMatches(resource, 'resource:update'));
};

export const isDeletable = (build: GraphileBuild.Build, resource: PgTableResource) => {
  if (resource.parameters) return false;
  if (!resource.codec.attributes) return false;
  if (resource.codec.polymorphism) return false;
  if (resource.codec.isAnonymous) return false;
  if (!resource.uniques || resource.uniques.length < 1) return false;
  return Boolean(build.behavior.pgResourceMatches(resource, 'resource:delete'));
};
