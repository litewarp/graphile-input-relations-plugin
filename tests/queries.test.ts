import {beforeAll, beforeEach, describe, expect, it} from 'bun:test';
import {existsSync, readdirSync} from 'node:fs';
import {resolve} from 'node:path';
import {PgSimplifyInflectionPreset} from '@graphile/simplify-inflection';
import {execute, hookArgs} from 'grafast';
import {type SchemaResult, makeSchema} from 'graphile-build';
import {type ExecutionArgs, type ExecutionResult, parse, validate} from 'graphql';
import type {Pool} from 'pg';
import {
  makePgService,
  makeWithPgClientViaPgClientAlreadyInTransaction,
} from 'postgraphile/adaptors/pg';
import {PostGraphileAmberPreset} from 'postgraphile/presets/amber';
import {RelationshipMutationsPreset} from '../src/index.ts';
import {withPgClient, withPgPool} from './helpers.ts';
import {printOrderedSchema} from './print-ordered-schema.ts';

const __dirname = import.meta.dir;
const schemasDir = resolve(__dirname, 'schemas');

const readTextFile = async (path: string) => {
  const file = Bun.file(path);
  return await file.text();
};

const getSchemaPath = (schema: string, file?: string) =>
  file ? resolve(schemasDir, schema, file) : resolve(schemasDir, schema);

const getFixturesDir = (schema: string) =>
  resolve(schemasDir, schema, 'fixtures', 'queries');

const getAllFixtures = (schema: string) =>
  existsSync(getFixturesDir(schema)) ? readdirSync(getFixturesDir(schema)).sort() : [];

const readSchema = async (schema: string, file?: string) =>
  readTextFile(getSchemaPath(schema, file));

const readFixture = async (schema: string, fixture: string) =>
  readTextFile(resolve(getFixturesDir(schema), fixture));

const getSqlSchemas = () => readdirSync(schemasDir).sort();

const createPostGraphileSchema = async (pgPool: Pool, sqlSchema: string) => {
  const gs = await makeSchema({
    extends: [
      PostGraphileAmberPreset,
      PgSimplifyInflectionPreset,
      RelationshipMutationsPreset,
    ],
    pgServices: [
      makePgService({
        pool: pgPool,
        schemas: [sqlSchema],
      }),
    ],
  });
  await Bun.write(`./.tmp/${sqlSchema}.graphql`, printOrderedSchema(gs.schema));
  return gs;
};

const sqlSchemas = getSqlSchemas();

beforeAll(() => {
  // Ensure process.env.TEST_DATABASE_URL is set
  if (!process.env['TEST_DATABASE_URL']) {
    console.error(
      'ERROR: No test database configured; aborting. To resolve this, ensure environmental variable TEST_DATABASE_URL is set.'
    );
    process.exit(1);
  }
});

for (const sqlSchema of sqlSchemas) {
  describe(sqlSchema, () => {
    let gqlSchema: SchemaResult;

    beforeEach(async () => {
      // reset the database
      await withPgClient(async (tx) => {
        await tx.query(await readSchema(sqlSchema, 'schema.sql'));
        await tx.query(await readSchema(sqlSchema, 'data.sql'));
      });
      gqlSchema = await withPgPool(async (pool) =>
        createPostGraphileSchema(pool, sqlSchema)
      );
    });
    const fixtures = getAllFixtures(sqlSchema);
    if (fixtures.length > 0) {
      for (const fixture of fixtures) {
        it(`query=${fixture}`, async () => {
          const {schema, resolvedPreset} = gqlSchema;
          const query = await readFixture(sqlSchema, fixture);
          const document = parse(query);

          const errors = validate(schema, document);
          if (errors.length > 0) {
            throw new Error(
              `GraphQL validation errors:\n${errors.map((e) => e.message).join('\n')}`
            );
          }
          const args: ExecutionArgs = {
            schema,
            document,
          };
          await hookArgs(args, resolvedPreset, {});
          const result = await withPgClient<ExecutionResult>(async (pgClient) => {
            // We must override the context because we didn't use a pool above and so
            // we need to add our own client
            // NOTE: the withPgClient needed on context is **VERY DIFFERENT** to our
            // withPgClient test helper. We should rename our test helper ;)

            const contextWithPgClient = makeWithPgClientViaPgClientAlreadyInTransaction(
              pgClient,
              false
            );

            try {
              args.contextValue = {
                pgSettings: (args.contextValue as {pgSettings: unknown}).pgSettings,
                withPgClient: contextWithPgClient,
              };
              return (await execute(args)) as Promise<ExecutionResult>;
            } finally {
              await contextWithPgClient.release?.();
            }
          });
          if (result.errors) {
            console.log(result.errors.map((e) => e.originalError ?? e));
          }
          // console.log(JSON.stringify(result.data, null, 2));

          expect(result).toMatchSnapshot();
        });
      }
    }
  });
}
