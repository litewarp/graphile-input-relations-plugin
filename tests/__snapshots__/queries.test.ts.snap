// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`a > query=double-nested-update.graphql 1`] = `
{
  "data": {
    "updateChildById": {
      "child": {
        "dadParent": {
          "id": "WyJQYXJlbnQiLDJd",
          "parentName": "Samuel Alito",
        },
        "id": "WyJDaGlsZCIsMV0=",
        "name": "Johnny Cochran",
      },
    },
  },
}
`;

exports[`a > query=forward-connect-by-id.graphql 1`] = `
{
  "data": {
    "createChild": {
      "child": {
        "dadParent": {
          "id": "WyJQYXJlbnQiLDJd",
          "parentName": "dad",
        },
        "id": "WyJDaGlsZCIsM10=",
        "name": "My baby",
      },
    },
  },
}
`;

exports[`a > query=forward-connect-by-keys.graphql 1`] = `
{
  "data": {
    "createChild": {
      "child": {
        "id": "WyJDaGlsZCIsM10=",
        "momParent": {
          "id": "WyJQYXJlbnQiLDFd",
          "parentName": "mom",
        },
        "name": "My baby",
      },
    },
  },
}
`;

exports[`a > query=forward-create.graphql 1`] = `
{
  "data": {
    "createChild": {
      "child": {
        "dadParent": {
          "id": "WyJQYXJlbnQiLDNd",
          "parentName": "Jimmy Constraint",
        },
        "id": "WyJDaGlsZCIsM10=",
        "momParent": null,
        "name": "Johnny Q. Constraint",
      },
    },
  },
}
`;

exports[`a > query=forward-update-by-id.graphql 1`] = `
{
  "data": {
    "updateChildById": {
      "child": {
        "id": "WyJDaGlsZCIsMV0=",
        "momParent": {
          "id": "WyJQYXJlbnQiLDFd",
          "parentName": "Sally Ride",
        },
        "name": "Jimminy Cricket",
      },
    },
  },
}
`;

exports[`a > query=mixed-multi-level-insert.graphql 1`] = `
{
  "data": {
    "createChild": {
      "child": {
        "dadParent": {
          "childrenByDadParentId": {
            "nodes": [
              {
                "id": "WyJDaGlsZCIsM10=",
                "name": "Sandra Day",
              },
              {
                "id": "WyJDaGlsZCIsNF0=",
                "name": "George Washington",
              },
              {
                "id": "WyJDaGlsZCIsNV0=",
                "name": "Martha Washington",
              },
            ],
            "totalCount": 3,
          },
          "id": "WyJQYXJlbnQiLDNd",
          "parentName": "John Q. Washington",
        },
        "id": "WyJDaGlsZCIsM10=",
        "name": "Sandra Day",
      },
    },
  },
}
`;

exports[`a > query=reverse-connect-by-id.graphql 1`] = `
{
  "data": {
    "createParent": {
      "parent": {
        "childrenByDadParentId": {
          "edges": [
            {
              "node": {
                "id": "WyJDaGlsZCIsMl0=",
                "name": "child 2",
              },
            },
          ],
        },
        "id": "WyJQYXJlbnQiLDNd",
        "parentName": "Daddy-O",
      },
    },
  },
}
`;

exports[`a > query=reverse-connect-by-keys.graphql 1`] = `
{
  "data": {
    "createSchool": {
      "school": {
        "id": "WyJTY2hvb2wiLDRd",
        "name": "My School",
        "teachers": {
          "nodes": [
            {
              "id": "WyJUZWFjaGVyIiwyXQ==",
              "name": "teacher 2",
              "uniqueField": "random unique field 2",
            },
          ],
          "totalCount": 1,
        },
      },
    },
  },
}
`;

exports[`a > query=reverse-create.graphql 1`] = `
{
  "data": {
    "createParent": {
      "parent": {
        "childrenByDadParentId": {
          "nodes": [
            {
              "dadParent": {
                "id": "WyJQYXJlbnQiLDNd",
              },
              "id": "WyJDaGlsZCIsM10=",
              "momParent": null,
              "name": "child 1 of test f1",
            },
          ],
        },
        "childrenByMomParentId": {
          "nodes": [
            {
              "id": "WyJDaGlsZCIsNF0=",
              "momParent": {
                "id": "WyJQYXJlbnQiLDNd",
              },
              "name": "child 2 of test f1",
            },
            {
              "id": "WyJDaGlsZCIsNV0=",
              "momParent": {
                "id": "WyJQYXJlbnQiLDNd",
              },
              "name": "child 3 of test f1",
            },
          ],
        },
        "id": "WyJQYXJlbnQiLDNd",
        "parentName": "test f1",
      },
    },
  },
}
`;

exports[`a > query=reverse-update-by-keys.graphql 1`] = `
{
  "data": {
    "updateSchoolById": {
      "school": {
        "id": "WyJTY2hvb2wiLDJd",
        "name": "Greendale Community College",
        "teachers": {
          "nodes": [
            {
              "id": "WyJUZWFjaGVyIiwxXQ==",
              "name": "Edna Krabappel",
              "uniqueField": "random unique field 1",
            },
            {
              "id": "WyJUZWFjaGVyIiwyXQ==",
              "name": "Seymour Skinner",
              "uniqueField": "random unique field 2",
            },
          ],
          "totalCount": 2,
        },
      },
    },
  },
}
`;

exports[`a > query=reverse-update-by-node.graphql 1`] = `
{
  "data": {
    "updateSchoolById": {
      "school": {
        "id": "WyJTY2hvb2wiLDJd",
        "name": "Greendale Community College",
        "teachers": {
          "nodes": [
            {
              "id": "WyJUZWFjaGVyIiwxXQ==",
              "name": "teacher 1",
              "uniqueField": "56789",
            },
            {
              "id": "WyJUZWFjaGVyIiwyXQ==",
              "name": "teacher 2",
              "uniqueField": "01234",
            },
          ],
          "totalCount": 2,
        },
      },
    },
  },
}
`;

exports[`a > query=triple-nested-connect-by-id.graphql 1`] = `
{
  "data": {
    "createChild": {
      "child": {
        "dadParent": {
          "id": "WyJQYXJlbnQiLDJd",
          "parentName": "dad",
        },
        "id": "WyJDaGlsZCIsM10=",
        "name": "Sonia Sotomayor",
        "studentsByStudentId": {
          "nodes": [
            {
              "id": "WyJTdHVkZW50IiwyLDNd",
              "school": {
                "id": "WyJTY2hvb2wiLDJd",
                "name": "pine",
              },
            },
          ],
          "totalCount": 1,
        },
      },
    },
  },
}
`;

exports[`a > query=triple-nested-create.graphql 1`] = `
{
  "data": {
    "createChild": {
      "child": {
        "id": "WyJDaGlsZCIsM10=",
        "momParent": {
          "childrenByMomParentId": {
            "nodes": [
              {
                "id": "WyJDaGlsZCIsM10=",
                "name": "Hugo Black",
                "studentsByStudentId": {
                  "nodes": [],
                  "totalCount": 0,
                },
              },
              {
                "id": "WyJDaGlsZCIsNF0=",
                "name": "John Roberts",
                "studentsByStudentId": {
                  "nodes": [
                    {
                      "id": "WyJTdHVkZW50IiwyLDRd",
                      "school": {
                        "id": "WyJTY2hvb2wiLDJd",
                      },
                      "schoolId": 2,
                      "studentId": 4,
                    },
                  ],
                  "totalCount": 1,
                },
              },
            ],
            "totalCount": 2,
          },
          "id": "WyJQYXJlbnQiLDNd",
          "parentName": "Sandra Day O'Connor",
        },
        "name": "Hugo Black",
      },
    },
  },
}
`;
